src/spi.rs

```diff
@@ -709,3 +804,132 @@ impl_spi!(SPI3: spi_host_device_t_SPI3_HOST);
 impl_spi_any_pins!(SPI2);
 #[cfg(not(esp32c3))]
 impl_spi_any_pins!(SPI3);
+
+use crate::gpio::{Level, Output};
+use crate::task::CriticalSection;
+use crate::delay::Ets;
+
+pub struct GuardedSpiDevice<'d, DRIVER: Borrow<SpiMasterDriver<'d>>> {
+    device: EspSpiDevice<'d, DRIVER>,
+    mutex: CriticalSection,
+}
+
+impl<'d, DRIVER> GuardedSpiDevice<'d, DRIVER>
+where
+    DRIVER: Borrow<SpiMasterDriver<'d>> + 'd,
+{
+    pub fn new(device: EspSpiDevice<'d, DRIVER>) -> Self {
+        Self {
+            device,
+            mutex: CriticalSection::new(),
+        }
+    }
+}
+pub struct EspSCSSpiDevice<'d, DEVICE, DRIVER: Borrow<SpiMasterDriver<'d>>> {
+    shared_device: DEVICE,
+    cs_pin: PinDriver<'d, AnyOutputPin, Output>,
+    pre_delay: Option<u32>,
+    post_delay: Option<u32>,
+    _p: PhantomData<&'d DRIVER>,
+}
+
+impl<'d, DEVICE, DRIVER> EspSCSSpiDevice<'d, DEVICE, DRIVER>
+where
+    DEVICE: Borrow<GuardedSpiDevice<'d, DRIVER>>,
+    DRIVER: Borrow<SpiMasterDriver<'d>>,
+{
+    pub fn new(
+        shared_device: DEVICE,
+        cs: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
+        cs_level: Level,
+    ) -> Result<Self, EspError> {
+        let cs_ref: PeripheralRef<AnyOutputPin> = cs.into_ref().map_into();
+        let mut cs_pin = PinDriver::output(cs_ref)?;
+        cs_pin.set_level(cs_level)?;
+        Ok(Self {
+            shared_device,
+            cs_pin,
+            _p: PhantomData,
+            pre_delay: None,
+            post_delay: None,
+        })
+    }
+    /// Add an aditional delay of x in uSeconds before transaction
+    /// between chip select and first clk out
+    pub fn cs_pre_delay(mut self, delay: u32) -> Self {
+        self.pre_delay = Some(delay);
+        self
+    }
+
+    /// Add an aditional delay of x in uSeconds after transaction
+    /// between last clk out and chip select
+    pub fn cs_post_delay(mut self, delay: u32) -> Self {
+        self.post_delay = Some(delay);
+        self
+    }
+
+
+    pub fn transaction<R, E>(
+        &mut self,
+        f: impl FnOnce(&mut SpiBusMasterDriver<'d>) -> Result<R, E>,
+    ) -> Result<R, E>
+    where
+        E: From<EspError>,
+    {
+        let shared_device: &GuardedSpiDevice<'d, DRIVER> = self.shared_device.borrow();
+        let _lock = shared_device.mutex.enter();
+        let device: &EspSpiDevice<'d, DRIVER> = shared_device.device.borrow();
+        self.cs_pin.toggle()?;
+        if let Some(delay) = self.pre_delay {
+            Ets::delay_us(delay);
+        }
+        let result = device.transaction(f)?;
```

Comment:
Using `?` here is NOK. You have to keep the `Result` without checking, then toggle the CS pin, and only then inspect the result with `?`.